app/src/main/java/ch/epfl/sdp/appart/database/local/LocalDatabase.java

```diff
@@ -0,0 +1,1055 @@
+        futureProfilePic.exceptionally(e -> {
+            futureSuccess.completeExceptionally(e);
+            return null;
+        });
+
+        futureSuccess.complete(null);
+    }
+
+    /**
+     * This function performs the writing of a complete ad into local storage
+     * . It will create a folder for the added user only if it doesn't
+     * already exists (in which case it is updated). It will do the same for
+     * the ad. It also performs the writing of all the images. The photos of
+     * ad, the panoramas of the ad and the profile picture of the user (if he
+     * doesn't uses the default one).
+     *
+     * @param adId           the id of the ad
+     * @param cardId         the id of the card
+     * @param ad             the ad
+     * @param user           the user who posted the ad
+     * @param adPhotos       list of bitmaps representing the photos of the ad
+     * @param panoramas      list of panoramas representing the panoramas of
+     *                       the ad
+     * @param loadProfilePic the function that will perform the writing on
+     *                       disk of the user's profile picture. It receives
+     *                       the path representing the location of the
+     *                       profile picture and returns a completable future
+     *                       to indicate if the operation succeeded or not.
+     * @return a completable future that indicates if the operation succeeded
+     * or not
+     */
+    public CompletableFuture<Void> writeCompleteAd(String adId,
+                                                   String cardId, Ad ad,
+                                                   User user,
+                                                   List<Bitmap> adPhotos,
+                                                   List<Bitmap> panoramas,
+                                                   Function<String,
+                                                           CompletableFuture<Void>> loadProfilePic) {
+
+        if (adId == null || cardId == null || ad == null || user == null || adPhotos == null || panoramas == null)
+            throw new IllegalArgumentException();
+
+        LocalDatabasePaths.cardID = cardId;
+        LocalDatabasePaths.userID = user.getUserId();
+
+
+        CompletableFuture<Void> futureSuccess = new CompletableFuture<>();
+        if (!createAdFolder(ad.getPhotosRefs().size(),
+                ad.getPanoramaReferences().size())) {
+            futureSuccess.completeExceptionally(new LocalDatabaseException(
+                    "Error while creating the ad folder !"));
+            return futureSuccess;
+        }
+
+
+        /*  I think we do not need to do something specific if the folder
+            already exists, we will just overwrite the files
+            Maybe we need to check for the number of images so that if the ad
+            has one image less than before we do not could keep a stale image
+            in the folder
```

Comment:
I am not sure I understand your question, I will describe how I thought of this.

So, let's say you have an ad that originally has 8 photos. They will have name Photo0 to Photo7. Now, let's say the user updates the ad and removes 4 photos. The useful photos are now Photo0 to Photo3. When we will write the photos, only Photo0 to Photo3 are going to be overwritten. So, my current system removes Photo4 to Photo7 to be sure we do not leave anything "unused".

This is done by this piece of code (I inserted an image because I didn't found a way nicely format the text):

![studio64_2021-05-13_19-59-25](https://user-images.githubusercontent.com/47746747/118166454-c6617f00-b425-11eb-88de-c94b5b06109d.png)

Reply:
Wouldn't be nice to store the time the ad was created in firebase and in the local db. And each time there is a modification on firebase it update the time. Then we simply do a "!=" over both time which would imply a staled ad ? No ?

Reply:
Moreover, at the startup, how to we check that the ads are up to date ? It would be great to store the time for this too.

Reply:
I see, I forgot you were overwriting the first photos!